server/src/internalClusterTest/java/org/opensearch/search/backpressure/SearchBackpressureIT.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    public void testSearchShardTaskCancellationWithHighCpu() throws InterruptedException {
+        ClusterUpdateSettingsRequest updateSettingsRequest = new ClusterUpdateSettingsRequest();
+        updateSettingsRequest.persistentSettings(Settings.builder().put(CpuUsageTracker.SETTING_CPU_TIME_MILLIS_THRESHOLD.getKey(), 1000));
+        assertAcked(client().admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        client().execute(TestTransportAction.ACTION, generateRequestWithHigCpuUsage(0.8, 10000), new ActionListener<>() {
+            @Override
+            public void onResponse(TestResponse testResponse) {
+                fail("SearchShardTask should have been cancelled");
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.info("testSearchShardTaskCancellationWithHighCpu onFailure e = " + e.getMessage());
+                assertEquals(TaskCancelledException.class, e.getClass());
+                assertTrue(e.getMessage().contains("cpu usage exceeded"));
+                latch.countDown();
+            }
+        });
+
+        latch.await();
+        resetSettingValue(CpuUsageTracker.SETTING_CPU_TIME_MILLIS_THRESHOLD.getKey());
+    }
+
+    public void testSearchCancellationWithBackpressureDisabled() throws InterruptedException {
+        ClusterUpdateSettingsRequest updateSettingsRequest = new ClusterUpdateSettingsRequest();
+        updateSettingsRequest.persistentSettings(Settings.builder().put(SearchBackpressureSettings.SETTING_MODE.getKey(), "monitor_only"));
+        assertAcked(client().admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        client().execute(TestTransportAction.ACTION, generateRequestWithHigCpuUsage(0.8, 10000), new ActionListener<>() {
+            @Override
+            public void onResponse(TestResponse testResponse) {
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("SearchShardTask shouldn't have cancelled for monitor_only mode");
+                latch.countDown();
+            }
+        });
+
+        latch.await();
+        resetSettingValue(SearchBackpressureSettings.SETTING_MODE.getKey());
+    }
+
+    /**
+     * Returns a {@link TestRequest} which will continue running for given milliseconds
+     * @param duration time in milliseconds
+     */
+    private TestRequest generateRequestWithHighElapsedTime(long duration) {
+        return new TestRequest(() -> {
+            try {
+                Thread.sleep(duration);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        });
+    }
```

Comment:
I don't think this is the intended implementation.

When such a request is executed, it will _always_ take `duration` time to complete even if the `SETTING_ELAPSED_TIME_MILLIS_THRESHOLD` is defined with a smaller value. For example, this test must be taking 8 seconds to complete even when the cut-off threshold is defined as 5 seconds.

We wish to interleave some 'computational work' with cancellation checks (similar to how search paths work), so that we can preempt a task in the middle of the execution rather than at the end of it.

```
TestTransportAction::doExecute would have this:

// Interleave doWork and cancellation check in a busy-wait loop (until cancellation or timeout).
do {
    doWork(request);
} while (task.isCancelled() == false && not timed out);
```

```
Then doWork could look something like:

private void doWork(TestRequest request) throws InterruptedException {
    // Do some expensive work here (run a tight loop to exhaust CPU, or allocate bytes to exhaust heap).
    // You can add some parameters to TestRequest to identify what kind of work you want to simulate here.
    switch (request.type) {
        case SLEEP_WAIT:
            Thread.sleep(100);
            break
        case FOO:
            // do something
            break
        case BAR:
            // do something else
            break
    }
}
```

Note that you don't even need a `duration` parameter. Just sleep for a small duration _at a time_ (100 ms), stop if the task is cancelled, else repeat the work (until timeout).

Reply:
Got it. Thanks for debugging and sharing the approach.